src/main/java/org/folio/circulation/services/EventPublishingService.java

```diff
@@ -0,0 +1,171 @@
+package org.folio.circulation.services;
+
+import static java.util.concurrent.CompletableFuture.completedFuture;
+import static org.folio.circulation.support.JsonPropertyWriter.write;
+import static org.folio.circulation.support.Result.succeeded;
+import static org.folio.rest.util.OkapiConnectionParams.OKAPI_TENANT_HEADER;
+
+import java.util.Map;
+import java.util.UUID;
+import java.util.concurrent.CompletableFuture;
+import java.util.stream.Collectors;
+
+import org.folio.circulation.domain.CheckInProcessRecords;
+import org.folio.circulation.domain.Loan;
+import org.folio.circulation.domain.LoanAndRelatedRecords;
+import org.folio.circulation.domain.LoanRepository;
+import org.folio.circulation.domain.RequestAndRelatedRecords;
+import org.folio.circulation.support.Clients;
+import org.folio.circulation.support.Result;
+import org.folio.rest.jaxrs.model.Event;
+import org.folio.rest.jaxrs.model.EventMetadata;
+import org.folio.rest.util.OkapiConnectionParams;
+import org.folio.util.pubsub.PubSubClientUtils;
+
+import io.vertx.core.Context;
+import io.vertx.core.json.JsonObject;
+import io.vertx.core.logging.Logger;
+import io.vertx.core.logging.LoggerFactory;
+import io.vertx.ext.web.RoutingContext;
+
+public class EventPublishingService {
+  private static final Logger logger = LoggerFactory.getLogger(EventPublishingService.class);
+
+  public static final String ITEM_CHECKED_OUT_EVENT_TYPE = "ITEM_CHECKED_OUT";
+  public static final String ITEM_CHECKED_IN_EVENT_TYPE = "ITEM_CHECKED_IN";
+  public static final String ITEM_DECLARED_LOST_EVENT_TYPE = "ITEM_DECLARED_LOST";
+  public static final String LOAN_DUE_DATE_UPDATED_EVENT_TYPE = "LOAN_DUE_DATE_UPDATED";
+
+  public static final String USER_ID_FIELD = "userId";
+  public static final String LOAN_ID_FIELD = "loanId";
+  public static final String DUE_DATE_FIELD = "dueDate";
+  public static final String RETURN_DATE_FIELD = "returnDate";
+  public static final String DUE_DATE_CHANGED_BY_RECALL_FIELD = "dueDateChangedByRecall";
+
+  private final Map<String, String> okapiHeaders;
+  private final Context vertxContext;
+
+  private Clients clients;
+
+  public EventPublishingService(RoutingContext routingContext) {
```

Comment:
This class seems to have multiple responsibilities, it both constructs the relevant event (sometimes by fetching extra information) and it publishes the event to pub-sub, could these responsibilities be split?

Reply:
Split to EventPublisher and PubSubPublishingService

Reply:
What is the difference between those two classes?

Reply:
EventPublisher constructs events payloads to pass to the PubSubPublishingService. It's called from the code that wants to publish an event. PubSubPublishingService is responsible for actual publishing to PubSub. The whole chain looks like "other code -> EventPublisher -> PubSubPublishingService".

Reply:
Thanks for explaining that